### PR TITLE
bridge: Enable disabling bridge interface

### PR DESCRIPTION
**Pull request by ormergi**
The new `disableContainerInterface` parameter is added to the bridge plugin to
enable setting the container interface state down.

When the parameter is enabled, the container interface (veth peer that is placed
at the container ns) remain down (i.e: disabled).
The bridge and host peer interfaces state are not affected by the parameter.

Notes for reviewers:
- Since IPAM logic involve various configurations including waiting for addresses
to be realized and setting the interface state UP, the new parameter cannot work
with IPAM.
In case both IPAM and DisableContainerInterface parameters are set, the bridge
plugin will raise an error.
- CNI versions 0.1.0 and 0.2.0 result expect single IP addresses, where newer versions expect multiple addresses (slice).
  When the new parameter is set, there will be no IP to report, thus it cannot work with older versions.
  In newer versions it works and an empty IP addresses slice will return.

Related Issue: https://github.com/containernetworking/plugins/issues/951

### CODE REVIEW

File: plugins/main/bridge/bridge.go [hunk 1]
```diff
@@ -708,7 +715,9 @@ func cmdAdd(args *skel.CmdArgs) error {
 		}
 
 		if idx == len(retries)-1 {
-			return fmt.Errorf("bridge port in error state: %s", hostVeth.Attrs().OperState)
+			if !n.DisableContainerInterface {
```
Comment (EdDev) [inline plugins/main/bridge/bridge.go:718]: It does not look to be relevant here.
The "check bridge port state" needs to be bypass if the flag is set to true.

**Review by EdDev — COMMENTED** [review]
Thanks!

File: plugins/main/bridge/bridge.go [hunk 2]
```diff
@@ -530,6 +531,10 @@ func cmdAdd(args *skel.CmdArgs) error {
 
 	isLayer3 := n.IPAM.Type != ""
 
+	if isLayer3 && n.DisableContainerInterface {
+		return fmt.Errorf("cannot use IPAM when container interface should be disabled")
```
[the remark this answers is not in the document]
Comment (ormergi) [reply plugins/main/bridge/bridge.go]: Done

File: plugins/main/bridge/bridge.go (hunk 1 above)
Comment (ormergi) [reply plugins/main/bridge/bridge.go]: Done

**Comment by ormergi** [comment]
[Changes](https://github.com/containernetworking/plugins/compare/034585e9352b00514e771ac69b0ab7b426720764..7b9327957804b3fd2523027771054e94d8bdc39b): fix linter issues

File: plugins/main/bridge/bridge.go [hunk 3]
```diff
@@ -530,6 +531,10 @@ func cmdAdd(args *skel.CmdArgs) error {
 
 	isLayer3 := n.IPAM.Type != ""
 
+	if isLayer3 && n.DisableContainerInterface {
+		return fmt.Errorf("cannot use IPAM when DisableContainerInterface flag is set")
```
Comment (EdDev) [inline plugins/main/bridge/bridge.go:535]: Can you check this scenario in the tests?

Comment (ormergi) [reply plugins/main/bridge/bridge.go:535]: Done

**Comment by ormergi** [comment]
[Changes](https://github.com/containernetworking/plugins/compare/a12ba15b964a342e8d9537a6e82f8197585ea9f9..7e131a0076de6bc69ed0ebc54c5903604ed940fa): failing test when both IPAM and disable-interface parameters are set

File: plugins/main/bridge/bridge_test.go [hunk 4]
```diff
@@ -2461,6 +2490,36 @@ var _ = Describe("bridge Operations", func() {
...
+
+		It(fmt.Sprintf("[%s] should set the container veth peer state down", ver), func() {
+			Expect(originalNS.Do(func(ns.NetNS) error {
+				defer GinkgoRecover()
+				tc := testCase{
+					cniVersion:       ver,
+					disableContIface: true,
+					isLayer2:         true,
+					AddErr020:        "cannot convert: no valid IP addresses",
+					AddErr010:        "cannot convert: no valid IP addresses",
```
Comment (AlonaKaplan) [inline plugins/main/bridge/bridge_test.go:2517]: Why are those errors expected?

Comment (ormergi) [reply plugins/main/bridge/bridge_test.go:2517]: CNI versions 0.1.0 and 0.2.0 result expect single IP addresses, where newer versions expect multiple addresses (slice).
When the new parameter is set, there will be no IP to report, thus it cannot work with older versions.
In newer versions it works and an empty IP addresses slice will return.

Comment (AlonaKaplan) [reply plugins/main/bridge/bridge_test.go:2517]: Can you please add this explanation to the PR message?

Comment (ormergi) [reply plugins/main/bridge/bridge_test.go:2517]: Done

**Review by mlguerrero12 — COMMENTED** [review]
have you checked if there is something to do for the check command? Does it verify that the link is up?

File: plugins/main/bridge/bridge.go [hunk 5]
```diff
@@ -692,23 +698,28 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+				break
+			}
 
-		if idx == len(retries)-1 {
-			return fmt.Errorf("bridge port in error state: %s", hostVeth.Attrs().OperState)
+			if idx == len(retries)-1 {
+				return fmt.Errorf("bridge port in error state: %s", hostVeth.Attrs().OperState)
+			}
 		}
 	}
```
[the remark this answers is not in the document]
Comment (ormergi) [reply plugins/main/bridge/bridge.go:724]: I have posted a PR for small refactoring around the places this PR change https://github.com/containernetworking/plugins/pull/998
Can we please address what you suggest on a follow up PR?

**Comment by ormergi** [comment]
> have you checked if there is something to do for the check command? Does it verify that the link is up?

Hi @mlguerrero12 thanks for the review, from what I saw the check cmd does not check the interface state at all.

File: plugins/main/bridge/bridge.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:724]: sure

**Review by mlguerrero12 — COMMENTED** [review]
LGTM, thanks @ormergi

**Comment by ormergi** [comment]
Hey @squeed please take another look, let me know if there is anything else we should do :slightly_smiling_face:

**Comment by ormergi** [comment]
Updating docs https://github.com/containernetworking/cni.dev/pull/137

**Review by squeed — APPROVED** [review]

**Comment by squeed** [comment]
Thanks for your contribution!

**Pull request merged**